File: 45-jump-game-ii/45-jump-game-ii.py
class Solution(object):
    def jump(self, nums):
        
        #Time Complexity = O(n)
        #Space Complexity = O(1)
        
        maxReach , steps , jump = nums[0] , nums[0] , 1
        
        if(len(nums) <= 1):
            return 0;
        
        if(nums[0] == 0):
            return 0
        
        for i in range(1 , len(nums)):
            if (i == len(nums) - 1):
                return jump
            maxReach = max(maxReach , i + nums[i])
            steps -= 1
            if (steps == 0):
                jump+=1
                steps = maxReach - i
                
        return jump
                
            
       
        

File: 45-jump-game-ii/45-jump-game-ii.cpp
class Solution {
public:
    int solve(int idx , vector<int>& nums , vector<int>&dp){
        
        if(idx  >= nums.size()-1) return 0;
        
        if(dp[idx] != -1) return dp[idx];
        
        int ans = 1e9;
        
        for(int i = 1 ; i <= nums[idx] ; i++){
            int jump = 1 + solve(idx + i , nums , dp);
            ans = min(ans , jump);
        }
        return dp[idx] = ans;
    }
    int jump(vector<int>& nums) {
        
        int n = nums.size();
        vector<int>dp(n , -1);
        int ans = solve(0 , nums ,dp);
        return ans == 1e9 ? 0 : ans;
        
    }
};